be/src/exec/pipeline/fragment_executor.cpp

```diff
@@ -107,8 +107,23 @@ Status FragmentExecutor::prepare(ExecEnv* exec_env, const TExecPlanFragmentParam
     LOG(INFO) << "Prepare(): query_id=" << print_id(query_id)
               << " fragment_instance_id=" << print_id(params.fragment_instance_id) << " backend_num=" << backend_num;
 
-    _fragment_ctx->set_runtime_state(
-            std::make_unique<RuntimeState>(query_id, fragment_instance_id, query_options, query_globals, exec_env));
+    int32_t degree_of_parallelism = 1;
+    if (request.__isset.pipeline_dop && request.pipeline_dop > 0) {
+        degree_of_parallelism = request.pipeline_dop;
+    } else {
+        // default dop is a half of the number of hardware threads.
+        degree_of_parallelism = std::max<int32_t>(1, std::thread::hardware_concurrency() / 2);
+    }
+
+    if (query_options.__isset.mem_limit) {
+        auto copy_query_options = query_options;
+        copy_query_options.mem_limit *= degree_of_parallelism;
```

Comment:
Why multiply by degree_of_parallelism ?

Reply:
Session variable `exec_mem_limit` means the largest bytes that one fragment instance can use.
For non-pipeline engine, it is ok
For pipeline engine, one fragment instance can have multipy pipeline drivers which share the `exec_mem_limit`

So why  multiply by `degree_of_parallelism`? Make sure that user change from non-pipeline engine to pipeline engine without modifying session variable `exec_mem_limit`

The meaning of `exec_mem_limit` becomes the largest bytes that the maximum number of bytes per degree of parallelism can use